pkg/riff/commands/requestprocessor_invoke_test.go

```diff
@@ -0,0 +1,408 @@
+						t.Errorf("expected command output to contain %q, actually %q", expected, output)
+					}
+				}
+			},
+		},
+		{
+			Name: "unknown ingress",
+			Args: []string{requestprocessorName},
+			GivenObjects: []runtime.Object{
+				requestProcessor,
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "istio-system",
+						Name:      "istio-ingressgateway",
+					},
+					Spec: corev1.ServiceSpec{
+						Ports: []corev1.ServicePort{},
+					},
+				},
+			},
+			ShouldError: true,
+		},
+		{
+			Name:       "missing ingress",
+			Args:       []string{requestprocessorName},
+			ExecHelper: "RequestProcessorInvoke",
+			GivenObjects: []runtime.Object{
+				requestProcessor,
+			},
+			ShouldError: true,
+		},
+		{
+			Name: "request processor not ready",
+			Args: []string{requestprocessorName},
+			GivenObjects: []runtime.Object{
+				&requestv1alpha1.RequestProcessor{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: defaultNamespace,
+						Name:      requestprocessorName,
+					},
+					Status: requestv1alpha1.RequestProcessorStatus{
+						Domain: fmt.Sprintf("%s.example.com", requestprocessorName),
+					},
+				},
+				ingressService,
+			},
+			ShouldError: true,
+		},
+		{
+			Name: "request processor missing domain",
+			Args: []string{requestprocessorName},
+			GivenObjects: []runtime.Object{
+				&requestv1alpha1.RequestProcessor{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: defaultNamespace,
+						Name:      requestprocessorName,
+					},
+					Status: requestv1alpha1.RequestProcessorStatus{
+						Status: duckv1alpha1.Status{
+							Conditions: []duckv1alpha1.Condition{
+								{Type: requestv1alpha1.RequestProcessorConditionReady, Status: "True"},
+							},
+						},
+					},
+				},
+				ingressService,
+			},
+			ShouldError: true,
+		},
+		{
+			Name: "missing request processor",
+			Args: []string{requestprocessorName},
+			GivenObjects: []runtime.Object{
+				ingressService,
+			},
+			ShouldError: true,
+		},
+		{
+			Name:       "curl error",
+			Args:       []string{requestprocessorName},
+			ExecHelper: "RequestProcessorInvokeError",
+			GivenObjects: []runtime.Object{
+				requestProcessor,
+				ingressService,
+			},
+			ShouldError: true,
+			Verify: func(t *testing.T, output string, err error) {
+				for _, expected := range []string{
+					"curl localhost -H 'Host: test-requestprocessor.example.com'\n",
+					"exit status 255\n",
+				} {
+					if !strings.Contains(output, expected) {
+						t.Errorf("expected command output to contain %q, actually %q", expected, output)
+					}
+				}
+			},
+		},
+	}
+
+	table.Run(t, commands.NewRequestProcessorInvokeCommand)
+}
+
+func TestHelperProcess_RequestProcessorInvoke(t *testing.T) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		return
+	}
+	// TODO assert arguments
+	fmt.Fprintf(os.Stderr, "Command executed: %s\n", shellquote.Join(argsAfterBareDoubleDash(os.Args)...))
+	os.Exit(0)
```

Comment:
Won't this terminate the test executable?

Reply:
yes, and it's intentional. Vanilla test runs will be caught by `if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {` and return before hitting the exit. When the test is run from within the exec call, it will hit the line and exit as normal.

The gotcha is that there are two levels of `go test` processes, this is the inner process. See the ExecHelper docs https://github.com/projectriff/riff/pull/1269/files#diff-ca1a901d21e48e7cd3b09c67650a643dR110 for detail and https://npf.io/2015/06/testing-exec-command/ for the inspiration.